Experimental SQL support via dfsql

btseytlin opened the pull request:
## What do these changes do?

Adds the ability to query Modin DataFrames with SQL statements.

How it works:
1. Install modin with [dfsql](https://github.com/mindsdb/dfsql).
``pip install modin[sql]``
2. Query using the `sql_query` interface:
```
>>> import modin.pandas as pd
>>> import modin.experimental.sql as mdsql
>>> df = pd.DataFrame({'col1': [1, 2, 3]})
>>> df.head()
   col1
0     1
1     2
2     3
>>> mdsql.sql_query("SELECT col1 FROM df WHERE col1 = 1", from_tables={'df': df})
1
``` 
3. Alternatively, use the added `.sql` DataFrame accessor. It's registered when `modin.experimental.sql` is imported. In this case the `FROM` can be omitted:
```
>>> df.sql("SELECT col1 WHERE col1 = 1")
UserWarning: `from_dict` defaulting to pandas implementation.
1
```

- [x] commit message follows format outlined [here](https://modin.readthedocs.io/en/latest/contributing.html)
- [x] passes `flake8 modin`
- [x] passes `black --check modin`
- [x] signed commit with `git commit -s` <!-- you can amend your commit with a signature via `git commit -amend -s` -->
- [x] Resolves #? <!-- issue must be created for each patch -->
- [x] tests added and passing

devin-petersohn commented on modin/experimental/sql/__init__.py:
```diff
@@ -0,0 +1,27 @@
+
+try:
+    import dfsql
+    from dfsql import sql_query
```
```suggestion
    from dfsql import sql_query as query
```

devin-petersohn commented on modin/experimental/sql/__init__.py:
```diff
@@ -0,0 +1,27 @@
+    )
+    raise
+
+__all__ = ["dfsql", "sql_query"]
```
```suggestion
__all__ = ["query"]
```

btseytlin replied:
I would rather leave `dfsql` exposed in case users want to use the full functonality of dfsql, like persisting tables

devin-petersohn replied:
Would you prefer `from modin.experimental.sql import dfsql` or `import dfsql`?

btseytlin replied:
Alright, it makes sense

devin-petersohn commented on modin/experimental/sql/__init__.py:
```diff
@@ -0,0 +1,27 @@
+import warnings
+
+try:
+    import dfsql
```
Will users use this import? 
```suggestion
```

devin-petersohn commented on modin/experimental/sql/__init__.py:
```diff
@@ -0,0 +1,27 @@
+    from dfsql import sql_query
+    import dfsql.extensions
+except ImportError:
+    warnings.warn(
```
I prefer to throw our own error message and not rethrow the `ImportError`. I think it might give too many error messages if we add an extra warning.

btseytlin replied:
I haven't found a fitting custom exception class in Modin. What should I use to throw the error?

devin-petersohn replied:
You can just raise an `ImportError`.

btseytlin replied:
Maybe remove the `try except` altogether then? Not very Pythonic to override an `ImportError` with custom text, if I understood your meaning correctly

devin-petersohn replied:
Let's just leave it as you have it. I think it will be ok.

devin-petersohn commented on setup.py:
```diff
@@ -41,8 +41,8 @@ def is_pure(self):
 spreadsheet_deps = ["modin-spreadsheet>=0.1.0"]
-
-all_deps = dask_deps + ray_deps + remote_deps + spreadsheet_deps
+sql_deps = ["dfsql==0.2.2"]
```
How quickly will this evolve? Is pinning to 0.2.2 a good idea or should we have a `>=` so we don't need to bump every release?

btseytlin replied:
I think you are right, should be a good idea to use `>=`

devin-petersohn reviewed (changes requested):
I created an issue for you to link in your PR: https://github.com/modin-project/modin/issues/2874. The commit message will need to have it as well. Something like:

```
FEAT-#2874: Experimental SQL support via dfsql
```
We should also add some sanity tests.

I can help you finish this up if you prefer.

btseytlin commented:
> I created an issue for you to link in your PR: #2874. The commit message will need to have it as well. Something like:
> 
> ```
> FEAT-#2874: Experimental SQL support via dfsql
> ```
> 
> We should also add some sanity tests.
> 
> I can help you finish this up if you prefer.

Added test, squashed commits and set the commit message

devin-petersohn reviewed (commented):
Looks good. Left one comment, let me know if it makes sense!

devin-petersohn reviewed (approved):
@btseytlin LGTM, thanks!
